generator/Can_Library_c.py

```diff
@@ -52,121 +56,132 @@ def write(output_path, spec_path, base_path):
-                            "_swap_value, bitstring, " + str(segment.position) + ", " + str(segment.length) + ");\n\n")
+            for msg in bus.messages:
+
+                # Write CAN_PACK
+                fw(
+                    'CAN_PACK(' + coord(bus.name, msg.name) + ') {' '\n'
+                    '\t' 'uint64_t bitstring = 0;' '\n'
+                )
+
+                for seg in msg.segments:
+                    if not msg.is_big_endian and (seg.c_type.startswith('int') or
+                                                  seg.c_type.startswith('uint')):
+                        fw(
+                            '\t' + seg.c_type + ' ' + seg.name + '_swap_value = swap_' + seg.c_type[:-2] + '(type_in->' + seg.name + ');' '\n'
+                            '\t' 'bitstring = INSERT(' + seg.name + '_swap_value, bitstring, ' + str(seg.position) + ', ' + str(seg.length) + ');' '\n\n'
+                        )
                     else:
-                        f.write(
-                            "  bitstring = INSERT(type_in->" + segment.name + ", bitstring, " + str(segment.position) +
-                            ", " + str(segment.length) + ");\n")
-
-                length = max(seg.position + seg.length for seg in message.segments)
-
-                f.write(
-                    "  from_bitstring(&bitstring, can_out->data);\n" +
-                    "  can_out->id = " + message.name.upper() + "__id;\n" +
-                    "  can_out->len = " + str(ceil(length / 8)) + ";\n" +
-                    "}\n\n")
-
-                # Write FROM_CAN
-                f.write(
-                    "FROM_CAN(Can_" + message.name + ") {\n" +
-                    "  uint64_t bitstring = 0;\n" +
-                    "  to_bitstring(can_in->data, &bitstring);\n")
-                for segment in message.segments:
-                    if segment.c_type == "enum":
-                        enum_name = "Can_" + message.name + "ID_T"
-
-                        f.write(
-                            "  type_out->" + segment.name + " = (" + enum_name + ")EXTRACT(bitstring, " +
-                            str(segment.position) + ", " + str(segment.length) + ");\n")
-                    elif segment.c_type == "bool":
-                        f.write(
-                            "  type_out->" + segment.name + " = EXTRACT(bitstring, " + str(segment.position) + ", " +
-                            str(segment.length) + ");\n")
+                        fw(
+                            '\t' 'bitstring = INSERT(type_in->' + seg.name + ', bitstring, ' + str(seg.position) +
+                            ', ' + str(seg.length) + ');' '\n'
+                        )
+
+                length = max(seg.position + seg.length for seg in msg.segments)
+
+                fw(
+                    '\t' 'from_bitstring(&bitstring, can_out->data);' '\n'
+                    '\t' 'can_out->id = ' + msg.name.upper() + '__id;' '\n'
```

Comment:
msg.name.upper() no longer valid (ex: VCUTODASH__id is no longer valid here)

Reply:
When Ctrl+Shitf+F fails you.
Press F to pay respects.